### PR TITLE
Initial wast support for components

### PR DESCRIPTION
**Pull request by sunfishcode**
This implements parsing for the [component text format](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md), and partial support for name resolution and encoding. There are several TODOs that I'll submit follow-up incremental PRs for once this base PR has landed.

### CODE REVIEW

File: crates/wast/tests/parse-component.rs
Comment (alexcrichton) [inline crates/wast/tests/parse-component.rs:67]: Could all of these tests be added to `tests/roundtrip.rs` and/or something like `tests/local/component-model`? They ideally would all exercise wasmparser's validation and wasmprinter's printing for round-trip-ness as well.

File: crates/wast/src/ast/alias.rs
Comment (alexcrichton) [inline crates/wast/src/ast/alias.rs:83]: Conventionally most of the parsers in `wast` delegate surrounding parenthesis parsing to the outer parsing context rather than the inner parsing context, so whomever parses aliases would parse these parens in theory.

Comment (alexcrichton) [reply crates/wast/src/ast/alias.rs:83]: (this is true for a number of the parsers I see below but I'll just comment here)

I should also point out that this isn't 100% true throughout `wast` but IIRC it's the style I tried to adhere to where possible

File: crates/wast/src/ast/token.rs
Comment (alexcrichton) [inline crates/wast/src/ast/token.rs:701]: I think in theory this shouldn't be necessary since the parser should otherwise report eof (or so I thought), but did the need for this arise?

Comment (alexcrichton) [inline crates/wast/src/ast/token.rs:701]: I think in theory this shouldn't be necessary since the parser should otherwise report eof (or so I thought), but did the need for this arise?

File: crates/wast/src/parser.rs
Comment (alexcrichton) [inline crates/wast/src/parser.rs:594]: Were these added opportunistically or was this needed for parsing somewhere internally? It wasn't my intention to really go infinitely long here with the peeking, and ideally we wouldn't add the desire to go too much further than the current spec needs.

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/component_names.rs:331]: This looks pretty similar to the `Namespace` helper for modules, so could that be reused?

File: crates/wast/src/resolve/deinline_import_export.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/deinline_import_export.rs:5]: How come the gensym bits went from something global to a struct-to-pass-around?

(IIRC I originally passed it around but it was such a pain to do so I stopped doing that)

**Review by alexcrichton — COMMENTED** [review]
Seems all pretty reasonable to me, but I think that we're going to want to disable something at compile-time or otherwise the fuzzers are likely going to get real angry at this. I think we may want a Cargo feature to control the internals of `Wat` parsing so the fuzzers don't try to fuzz all this (also so it's not exposed by default if we publish since it's in an incomplete state)

Otherwise the main stylistic comment is that most parsers are starting off by parsing their parens when I typically try where possible to have it the other way around and parsers don't parse their own parens and parents parse the outer parens of the item they're sub-parsing.

File: crates/wast/src/parser.rs
Comment (sunfishcode) [reply crates/wast/src/parser.rs]: It was used in earlier version of this PR, as the text format has some tricky parsing ambiguities, but it's not needed anymore. I've now removed it.

File: crates/wast/src/resolve/deinline_import_export.rs
Comment (sunfishcode) [reply crates/wast/src/resolve/deinline_import_export.rs:5]: That was just me trying to understand name resolution for nested modules and wanting to understand what state was kept where. I can reinstate the thread-local if it's clearer that way, but the current code doesn't need to pass it around that many places.

File: crates/wast/src/ast/alias.rs
Comment (sunfishcode) [reply crates/wast/src/ast/alias.rs:83]: I personally would like to replace manual paren-consuming and manual token-lookahead with an s-expression parser that produces a tree with strings at the leaves, and then parse that :-).

File: crates/wast/src/ast/token.rs
Comment (sunfishcode) [reply crates/wast/src/ast/token.rs:701]: It comes in handy when resolving some of the parsing ambiguities in the component-model text format.

File: crates/wast/src/ast/alias.rs
Comment (alexcrichton) [reply crates/wast/src/ast/alias.rs:83]: In the meantime would you be ok sticking to to general idioms of the rest of the library as-is?

File: crates/wast/src/ast/token.rs
Comment (alexcrichton) [reply crates/wast/src/ast/token.rs:701]: Are those locations where sub-parsers or eof don't work? (or places where the text format could be changed to avoid the need for such lookahead?)

File: crates/wast/src/ast/alias.rs
Comment (sunfishcode) [reply crates/wast/src/ast/alias.rs]: I've now moved most of the parse implementations to this style. The type rules still consume the paren though, because intertype doesn't always start with a paren, and it seems simpler to have it optionally parse a paren than to split the parsing into two places.

File: crates/wast/src/ast/token.rs
Comment (sunfishcode) [reply crates/wast/src/ast/token.rs]: I've now reorganized the parsing to avoid the need for this, so `RParen` is now removed.

File: crates/wast/tests/parse-component.rs
Comment (sunfishcode) [reply crates/wast/tests/parse-component.rs:67]: Yes, that will be good to do once everything roundtrips, but wasmparser doesn't appear to support everything that this new wast encoder is encoding yet. I'll move things to roundtrip tests once we work these out.

File: crates/wast/src/resolve/mod.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/mod.rs:113]: I think one thing this is missing is the de-inlining/elaboration, right? 

Or, more generally, would you be up for making a checklist of remaining TODO items after this?

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/component_names.rs:17]: This option seems to be passed around to most functions here, along with the redefinitions of `'a` and `'b`. Would you be up for using a style similar to the core module resolver where a `struct` is used to contain this `Vec`?

Also reaidng over this I'm not sure if the `'b` lifetime parameter throughout this module is still necessary, although I may be missing where it's needed. At least for this function though it's not required.

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/component_names.rs:661]: I think this can return a normal resolve error here perhaps?

File: crates/wast/src/binary.rs
Comment (alexcrichton) [reply crates/wast/src/binary.rs:861]: Or also in general could the `eprintln` be removed from this module?

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/component_names.rs:539]: This pattern of push, loop, add aliases, etc shows up in a few places, could this be wrapped up into a helper method to avoid duplicating the logic to insert aliases based on the results of resolution?

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [inline crates/wast/src/resolve/component_names.rs:478]: Is it necessary to maintain these index spaces for component types? I thought that the only thing that mattered at this layer was names-of-types?

**Review by alexcrichton — APPROVED** [review]
Looking reasonable to me! One thing I'm not certain about is what's remaining after this, do you feel like you've got a handle on that to write up an issue with remaining work-items?

**Comment by alexcrichton** [comment]
I'm going to be helping to push this over the finish line now and for now I plan on continuing to wean down the size of this by splitting out refactorings such as https://github.com/bytecodealliance/wasm-tools/pull/577. I've done an initial merge of this and that branch to the point where this builds, but there's still quite a lot of `todo!`, `TODO`, `eprintln`, `panic!`, and features that need implementing here before I finish landing this. My hope is to get this to a point which is land-able in the near future which is well-tested too (e.g. roundtrip testing) and then work on subsequent features and sugar required by the text format.

**Comment by alexcrichton** [comment]
Ok I have made a bunch more progress on this but there is yet more still to be made. The branch is in a pretty good place passing a good chunk of tests and most of the remaining failing tests are failing because they need updates from module linking. My current plan with this PR is:

* Probably don't land this behind a feature gate, just land it now.
* Split out more minor refactorings and bugfixes I've been accmulating here and land separately
* Postpone refactoring the `resolve.rs` module as per [this comment](https://github.com/bytecodealliance/wasm-tools/pull/529#discussion_r858787833) and [this comment](https://github.com/bytecodealliance/wasm-tools/pull/529#discussion_r858805377)
* Postpone fleshing out tests beyond what's already included


Beyond all that this PR is, uh, still quite chonky at 6k+ lines. I feel like a big chunk is just the verbosity of the `wast` crate itself, but the component model is also quite large and encompasses a lot (and everything here is brand new, not much ended up being able to be shared).

I will re-review the size of this PR once I get all the refactorings out and see if we can't slim it down. But that's where I'm at! (won't be back at this again until Monday)

**Comment by alexcrichton** [comment]
Oh also the history via git here is an absolute mess now thanks to my endless merging of `main`, my plan is to just squash this all down when merging.

**Comment by alexcrichton** [comment]
Ok this is the final state and I think this is ready for merging. @peterhuene or @fitzgen did one of y'all want to take a look at this? I reviewed Dan's original work here but I've made a number of nontrivial changes here myself. Unfortunately I can't really easily generate a diff between what I last reviewed personally and what's here now due to all the internal refactorings.

Overall I don't think that we need to hold `wast` to too-high of a standard of correctness (as opposed to `wasmparser`) since it's primarily intended for developers. In that sense if y'all don't have time to look at this I think it's ok to land and fixup later with more tests and such.

My next follow-up to this will be to use coverage information to find holes in what we're not parsing and add test for that, followed by holes in what we're not validating and adding tests for that as well.

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [reply crates/wast/src/resolve/component_names.rs]: I'm splitting this out to https://github.com/bytecodealliance/wasm-tools/issues/599

File: crates/wast/src/resolve/component_names.rs
Comment (alexcrichton) [reply crates/wast/src/resolve/component_names.rs]: This is split out to https://github.com/bytecodealliance/wasm-tools/issues/599

File: crates/wast/src/component/module.rs
Comment (peterhuene) [inline crates/wast/src/component/module.rs:7]:
```suggestion
/// A nested WebAssembly module to be created as part of a component.
```

File: crates/wast/src/parser.rs
Comment (peterhuene) [inline crates/wast/src/parser.rs:274]: Am I misreading this or is this function identical to `peek2`?

**Review by peterhuene — COMMENTED** [review]
Looks good, just a few comments :+1:. I skipped over most of the test case changes.

Definitely don't need to block on my feedback since we're about to change things here quite a bit with the upcoming text/binary format changes.

File: crates/wast/src/component/binary.rs
Comment (alexcrichton) [reply crates/wast/src/component/binary.rs:222]: Yeah I think this will all need refactoring with the new index space splits, and I'm hoping that the index space split will lend a natural structure to the AST here once we implement it. There's a whole mess of test cases that should be passing but are failing tagged with #588 which I hope to get back to once we redo the binary format.

File: crates/wast/src/component/deftype.rs
Comment (alexcrichton) [reply crates/wast/src/component/deftype.rs:186]: Aha yeah this is actually something that turned up really quickly in my "let's use coverage to figure out what's not being tested" and it turns out this wasn't being tested in `wast`. I'll leave this as-is for now but will follow-up shortly with a refactoring to bring this in line with the model's expectation of encodings.

File: crates/wast/src/parser.rs
Comment (alexcrichton) [reply crates/wast/src/parser.rs]: Ah turns out this was dead code!

**Comment by alexcrichton** [comment]
Ok I'm going to go ahead and merge this. Thanks again @sunfishcode for all the initial work here and thanks @peterhuene for the review and putting up with my issues that are sometimes overeager!

File: crates/wast/src/component/deftype.rs
Comment (alexcrichton) [reply crates/wast/src/component/deftype.rs:186]: (specifically I'm changing this in https://github.com/bytecodealliance/wasm-tools/pull/602)

**Pull request merged**
